src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs

```diff
@@ -0,0 +1,250 @@
+
+                static bool compareIdentifiers(VariableIdentifier id1, VariableIdentifier id2) =>
+                    (!id1.Exists && !id2.Exists) || id1.Equals(id2);
+
+#if DEBUG
+                static void verifyRemainderNonExistant(VariableIdentifier[] variablesBySlot, int startIndex)
+                {
+                    for (int i = startIndex; i < variablesBySlot.Length; i++)
+                    {
+                        Debug.Assert(!variablesBySlot[i].Exists);
+                    }
+                }
+#endif
+            }
+
+            internal void SnapshotWalker(BoundNode node)
+            {
+                if (_snapshotMap != null)
+                {
+                    _snapshotMap[node] = Snapshot.SnapshotWalker(_walkerToSnapshot);
+                }
+            }
+
+            internal ImmutableDictionary<BoundNode, Snapshot> Snapshots => _snapshotMap.ToImmutableDictionary();
+
+            internal bool IncrementalSnapshotter => _snapshotMap != null;
+        }
+
+        /// <summary>
+        /// Used to copy variable slots and types from the NullableWalker for the containing method
+        /// or lambda to the NullableWalker created for a nested lambda or local function.
+        /// </summary>
+        internal sealed class VariableState
+        {
+            // Consider referencing the collections directly from the original NullableWalker
+            // rather than copying the collections. (Items are added to the collections
+            // but never replaced so the collections are lazily populated but otherwise immutable.)
+            internal readonly ImmutableDictionary<VariableIdentifier, int> VariableSlot;
+            internal readonly ImmutableArray<VariableIdentifier> VariableBySlot;
+            internal readonly ImmutableDictionary<Symbol, TypeWithAnnotations> VariableTypes;
+
+            // The nullable state of all variables captured at the point where the function or lambda appeared.
+            internal readonly LocalState VariableNullableStates;
+
+            internal VariableState(
+                ImmutableDictionary<VariableIdentifier, int> variableSlot,
+                ImmutableArray<VariableIdentifier> variableBySlot,
+                ImmutableDictionary<Symbol, TypeWithAnnotations> variableTypes,
+                LocalState variableNullableStates)
+            {
+                VariableSlot = variableSlot;
+                VariableBySlot = variableBySlot;
+                VariableTypes = variableTypes;
+                VariableNullableStates = variableNullableStates;
+            }
+        }
+
+        /// <summary>
+        /// Contains a checkpoint for the NullableWalker at any given point of execution, used for restoring the walker to
+        /// a specific point for speculatively analyzing a piece of code that does not appear in the original tree.
+        /// </summary>
+        internal readonly struct Snapshot
```

Comment:
>Snapshot [](start = 33, length = 8)

It feels like this structure can be significantly trimmed as we discussed offline.